### PR TITLE
functors task (1) - Kostiuchenko

### PR DESCRIPTION
**Pull request by kosnko**

### CODE REVIEW

File: functors_task_1/student/include/impl.h [hunk 1]
```diff
@@ -8,6 +8,21 @@
  */
 struct BeerOrganizer
 {
+    BeerOrganizer()
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:11]: Use in-place initialization
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c45-dont-define-a-default-constructor-that-only-initializes-data-members-use-in-class-member-initializers-instead

File: functors_task_1/student/include/impl.h [hunk 2]
```diff
@@ -8,6 +8,21 @@
...
+        : state{BeerBrand::None}
+    {}
+    BeerBrand operator()()
+    {
+        int nextIndex = getNextIndex();
+        state = (0 == nextIndex) ? BeerBrand::HoeGaarden : static_cast<BeerBrand>(nextIndex);
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:17]: please specify range with static constexpr variablies: "to" and "from"

File: functors_task_1/student/include/impl.h [hunk 3]
```diff
@@ -8,6 +8,21 @@
...
+        return state;
+    }
+private:
+    int getNextIndex()
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:21]: mark as const

File: functors_task_1/student/include/impl.h [hunk 4]
```diff
@@ -16,16 +31,35 @@ struct BeerOrganizer
...
+bool isExpensiveBeer(BeerBrand brand)
 {
+    switch (brand)
+    {
+        case BeerBrand::Corona:
+        case BeerBrand::HoeGaarden:
+        {
+            return true;
+        }
+        case BeerBrand::Leffe:
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:43]: Why do you need to mention all of these types?

File: functors_task_1/student/include/impl.h [hunk 5]
```diff
@@ -16,16 +31,35 @@ struct BeerOrganizer
...
+        }
+    }
 }
 
 /**
  * @todo Implement lambda beer country equality comparator
  * @return true if beer county is the same, false otherwise
  */
-auto sameCountry = [](/**???*/)
+auto sameCountry = [](BeerBrand brand1, BeerBrand brand2)
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:60]: _mark as const_

File: functors_task_1/student/include/impl.h [hunk 6]
```diff
@@ -40,9 +74,25 @@ struct MixingPolicy
      * Whiskey + SevenUp = SevenPlusSeven;
      * Others + Others = Oops;
      */
-    static Cocktail mix(/**???*/)
+    static Cocktail mix(AlcoholDrink alcDrink, NonAlcoholDrink nonAlcDrink)
     {
+        if(alcDrink == AlcoholDrink::Gin && nonAlcDrink == NonAlcoholDrink::LimeJuice)
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:79]: Please prefer switch

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: functors_task_1/student/include/impl.h (hunk 6 above)
Comment (kosnko) [reply functors_task_1/student/include/impl.h:79]: Done

File: functors_task_1/student/include/impl.h (hunk 4 above)
Comment (kosnko) [reply functors_task_1/student/include/impl.h:43]: corrected

File: functors_task_1/student/include/impl.h (hunk 1 above)
Comment (kosnko) [reply functors_task_1/student/include/impl.h:11]: Done

Comment (smay1613) [reply functors_task_1/student/include/impl.h:11]: did you pushed your changes?

Comment (kosnko) [reply functors_task_1/student/include/impl.h]: Already pushed

File: functors_task_1/student/include/impl.h (hunk 2 above)
Comment (kosnko) [reply functors_task_1/student/include/impl.h]: confused but done

File: functors_task_1/student/include/impl.h (hunk 3 above)
Comment (kosnko) [reply functors_task_1/student/include/impl.h]: done

File: functors_task_1/student/include/impl.h (hunk 5 above)
Comment (kosnko) [reply functors_task_1/student/include/impl.h]: done

File: functors_task_1/student/include/impl.h [hunk 7]
```diff
@@ -8,6 +8,20 @@
  */
 struct BeerOrganizer
 {
+    BeerBrand operator()()
+    {
+        int nextIndex = getNextIndex();
+        static constexpr int from = static_cast<int>(BeerBrand::HoeGaarden);
+        static constexpr int to = static_cast<int>(BeerBrand::Leffe);
+        state = (from <= nextIndex && to >= nextIndex) ? static_cast<BeerBrand>(nextIndex) : BeerBrand::HoeGaarden;
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:16]: why "from" and "to" are not used?

File: functors_task_1/student/include/impl.h [hunk 8]
```diff
@@ -16,16 +30,29 @@ struct BeerOrganizer
...
  */
-bool isExpensiveBeer(/**???*/)
+bool isExpensiveBeer(BeerBrand brand)
 {
+    switch (brand)
+    {
+        case BeerBrand::Corona:
+        case BeerBrand::HoeGaarden:
+        {
+            return true;
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:40]: avoid explicit true/false in your code, also just a plain logical expression is a best fit here

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: functors_task_1/student/include/impl.h [hunk 9]
```diff
@@ -40,9 +67,44 @@ struct MixingPolicy
...
+                    {
+                        return Cocktail::Gimlet;
+                    }
+                    case NonAlcoholDrink::GrapefruitJuice:
+                    {
+                        return Cocktail::Greyhount;
+                    }
+                    default:
+                    {
+                        return Cocktail::Oops;
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:87]: break and just go to a fallback at the end of the function

**Pull request closed without merging**